### PR TITLE
Query implementation for categories [Date: 03/01/2023] [Time: 21:35:22]

### PR DESCRIPTION
**Pull request by i-api**

### CODE REVIEW

**Comment by ChrisWilhelm** [comment]
@i-api This is not the required query! You need to query for all categories that are currently in the warehouse, not all categories.

**Review by ChrisWilhelm — CHANGES_REQUESTED** [review]
This is effectively the same query you had before. It just returns all categories, since a pallot will never be removed from the DB as we need to track the total weight and history. 

To check if something is in the warehouse, you need to check the field inWarehouse and if it is true or false. This should be included in the findMany query for pallots. Additionally, there is no need to query for all the categories initially. Instead, create a set of categoryIds (like you are currently doing), and then query like so: 

`categories = await prisma.findMany({where : {id: { in: categoryIds}, }, });`

File: server/src/controller/categoriesQuery.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  if (validate(req, res)) {
+    return res;
+  }
+
+  const [pallots, categories] = await Promise.all([prisma.pallot.findMany(), prisma.category.findMany()]);
+
+  let categoryIds_set = new Set(); // no duplicates
+  for (const pallot of pallots) {
+    for (const categoryIds of pallot.categoryIds) {
+      for (const categoryId of categoryIds) {
```
Comment (ChrisWilhelm) [inline server/src/controller/categoriesQuery.js:19]: also you have an extra for loop here

**Comment by ChrisWilhelm** [comment]
Also, @i-api, all of these new routes should go into the existing routes and controller files. We do not need new ones.

File: server/src/controller/categoriesQuery.js (hunk 1 above)
Comment (i-api) [reply server/src/controller/categoriesQuery.js]: I believe that categoryIds is of type Int[], so I think the for loop is needed? I might not be understanding this correctly, pls lmk

Screen Shot 2023-03-03 at 1 01 20 PM

**Comment by i-api** [comment]
> Also, @i-api, all of these new routes should go into the existing routes and controller files. We do not need new ones.

6f106350539b620abaa2dd61af0595ba4cdedc80

File: server/src/controller/categoriesQuery.js (hunk 1 above)
Comment (i-api) [reply server/src/controller/categoriesQuery.js]: Are you saying that each "pallot" has only one "categoryIds"?

Comment (i-api) [reply server/src/controller/categoriesQuery.js]: I think I understand now

**Comment by i-api** [comment]
> This is effectively the same query you had before. It just returns all categories, since a pallot will never be removed from the DB as we need to track the total weight and history.
> 
> To check if something is in the warehouse, you need to check the field inWarehouse and if it is true or false. This should be included in the findMany query for pallots. Additionally, there is no need to query for all the categories initially. Instead, create a set of categoryIds (like you are currently doing), and then query like so:
> 
> `categories = await prisma.findMany({where : {id: { in: categoryIds}, }, });`

8eeca286163aca440d812697eed1f9e78920ae92
45dca04b61898feffcb2eed70229b24e92baec59

File: server/src/controller/categoriesQuery.js (hunk 1 above)
Comment (i-api) [reply server/src/controller/categoriesQuery.js]: used flatMap() btw for better readability

**Review by jstelnik — APPROVED** [review]

**Review by ChrisWilhelm — APPROVED** [review]
LGTM

**Review by krutalp — APPROVED** [review]

**Comment by i-api** [comment]
> LGTM

Cool.

**Pull request merged**